### DISCUSSION TITLE
Unexpected token import.meta.env when running Cypress with Vite

### DISCUSSION THREAD

**Discussion by kcubero27** [cypress-io/cypress#25877] [Questions and Help]
Hello 👋🏼 
I'm trying to set up Cypress on a Vite project (together). However, I am not able to run Cypress e2e tests with the app: either using the output of the build neither the dev mode. The error is the following one:
```
Error: Webpack Compilation Error
./src/api/routes.ts 5:63
Module parse failed: Unexpected token (5:63)
File was processed with these loaders:
 * ../../Library/Caches/Cypress/12.6.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/ts-loader/index.js
You may need an additional loader to handle the result of these loaders.
| exports.API_ROUTES = void 0;
| exports.API_ROUTES = {
>     API_ROUTE_GET_PEOPLE: function () { return "".concat(import.meta.env.VITE_STAR_WARS_API_URL, "/people"); },
| };
| 
 @ ./cypress/e2e/people.cy.ts 5:15-46
```

Does anyone know how to fix this?  I would really appreciate it.

In order to reproduce it, you can go to https://github.com/kcubero27/react-code-challenge-template/tree/feat/configure-e2e-tests-%2316 and run `npm run dev` as well as `npm run test:e2e:open`. As soon as you try to run the people.cy.ts test, you'll start seeing the error.

**Comment by portyu9** [comment]
Hi @kcubero27 

The problem is that your E2E spec imports application code containing:

```js
import.meta.env.VITE_STAR_WARS_API_URL
```

but that imported module is being compiled by **Cypress's E2E Webpack preprocessor**, not by Vite. The stack trace confirms this with `Webpack Compilation Error` and `ts-loader`.

`import.meta.env` is transformed/provided by Vite, so simply running the application itself with Vite does not make the same transform available when Cypress bundles the E2E spec and everything that spec imports.

The cleanest solution is usually to avoid importing Vite-bound application modules into an E2E spec. For example, move route names/constants that both sides need into a module that does not read `import.meta.env`, and let the application combine them with its environment configuration.

If your E2E specs genuinely need Vite to process imported application source, you need a Vite-aware `file:preprocessor`. Cypress still defaults to Webpack for E2E; community preprocessors such as `cypress-vite` can be registered explicitly:

```js
import { defineConfig } from 'cypress'
import vitePreprocessor from 'cypress-vite'

export default defineConfig({
  e2e: {
    setupNodeEvents(on) {
      on('file:preprocessor', vitePreprocessor())
    },
  },
})
```

That is a custom/community E2E preprocessor, not Cypress's built-in Component Testing Vite bundler.

So the key point is: **your application uses Vite, but your E2E spec dependency graph is currently being compiled with Webpack.**

Please mark this answer as accepted if it helped, thank you.
